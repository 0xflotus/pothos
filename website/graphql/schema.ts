import { docsIndex } from '../util/build-index.mjs';
import { builder } from './builder';
import {
  DocsIndex,
  DocsSection,
  SearchMatch,
  SearchMatcher,
  SearchMatchKind,
  SearchResult,
} from './search-matcher';

const docsIndexFiles = docsIndex as DocsIndex[];

const DocsPage = builder.objectRef<DocsIndex>('DocsPage').implement({
  fields: (t) => ({
    title: t.exposeString('name'),
    description: t.exposeString('description'),
    link: t.exposeString('link'),
    md: t.exposeString('md'),
    html: t.exposeString('html'),
  }),
});

builder.enumType(SearchMatchKind, { name: 'SearchMatchKind' });

const SearchMatchRef = builder.objectRef<SearchMatch>('SearchMatch').implement({
  fields: (t) => ({
    kind: t.expose('kind', { type: SearchMatchKind }),
    section: t.expose('section', { type: DocsSectionRef, nullable: true }),
  }),
});

const SearchResultRef = builder.objectRef<SearchResult>('SearchResult').implement({
  fields: (t) => ({
    matches: t.expose('matches', { type: [SearchMatchRef] }),
    doc: t.expose('doc', { type: DocsPage, nullable: true }),
  }),
});

const DocsSectionRef = builder.objectRef<DocsSection>('DocsSection').implement({
  fields: (t) => ({
    heading: t.exposeString('heading'),
    link: t.exposeString('link'),
    md: t.exposeString('md'),
    html: t.exposeString('html'),
  }),
});

builder.queryType({
  fields: (t) => ({
    search: t.field({
      type: [SearchResultRef],
      args: {
        query: t.arg.string({
          required: true,
        }),
      },
      resolve: (_, { query }) => new SearchMatcher(query).matchDocs(docsIndexFiles).getResults(),
    }),
  }),
});

export const schema = builder.toSchema({});
