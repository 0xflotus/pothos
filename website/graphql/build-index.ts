/* eslint-disable unicorn/no-process-exit */
/* eslint-disable no-process-exit */
/* eslint-disable @typescript-eslint/no-implicit-any-catch */
// import algoliasearch from 'algoliasearch/lite';
import dotenv from 'dotenv';

try {
  dotenv.config();

  if (!process.env.NEXT_PUBLIC_ALGOLIA_APP_ID) {
    throw new Error('NEXT_PUBLIC_ALGOLIA_APP_ID is not defined');
  }

  if (!process.env.ALGOLIA_SEARCH_ADMIN_KEY) {
    throw new Error('ALGOLIA_SEARCH_ADMIN_KEY is not defined');
  }
} catch (error) {
  console.error(error);
  process.exit(1);
}

// const searchFormat = loadedFiles.map((file) => {
//   if (!(file.data.title as string)) {
//     throw new Error(`Missing title for ${file.path}`);
//   }

//   return {
//     objectID: file.path,
//     title: file.data.title as string,
//     description: file.data.description as string,
//     slug: (file.data.slug as string) || toSlug(file.path),
//     tagsCollection: { tags: file.data.tags as string[] },
//     date: file.data.date as string,
//     type: 'docs',
//     content: file.content,
//   };
// });
