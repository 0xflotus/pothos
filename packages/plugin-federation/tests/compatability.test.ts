import { printSchema } from 'graphql';
import { printSubgraphSchema } from '@apollo/subgraph';
import { schema } from './example/compatibility/products/schema';

describe('federation', () => {
  describe('compatibility', () => {
    it('generates expected schema', () => {
      expect(printSubgraphSchema(schema)).toMatchSnapshot();
      expect(printSchema(schema)).toMatchSnapshot();
    });
  });
});
