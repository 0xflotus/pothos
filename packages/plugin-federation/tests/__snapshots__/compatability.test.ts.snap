// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`federation compatibility generates expected schema 1`] = `
"directive @inaccessible on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION

directive @link(import: [link__Import], url: String!) on SCHEMA

directive @override(from: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

type Product
  @key(fields: \\"sku package\\")
  @key(fields: \\"sku variation { id }\\")
{
  id: ID!
  sku: String
  package: String
  dimensions: ProductDimension
  variation: ProductVariation
  notes: [String!] @inaccessible @tag(name: \\"internal\\")
  createdBy: User @provides(fields: \\"totalProductsCreated\\")
}

type ProductDimension
  @shareable
{
  size: String
  weight: Int
  unit: String @inaccessible
}

type ProductVariation {
  id: ID!
}

type Query {
  produce(id: ID!): Product
}

type User
  @key(fields: \\"email\\")
  @extends
{
  email: String! @external
  totalProductsCreated: Int @external
  name: String @override(from: \\"users\\")
}

scalar link__Import"
`;

exports[`federation compatibility generates expected schema 2`] = `
"directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @inaccessible on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION

directive @key(fields: String!, resolvable: Boolean = true) repeatable on INTERFACE | OBJECT

directive @link(import: [link__Import], url: String!) on SCHEMA

directive @override(from: String!) on FIELD_DEFINITION

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

directive @tag(name: String!) repeatable on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION

type Product {
  createdBy: User
  dimensions: ProductDimension
  id: ID!
  notes: [String!]
  package: String
  sku: String
  variation: ProductVariation
}

type ProductDimension {
  size: String
  unit: String
  weight: Int
}

type ProductVariation {
  id: ID!
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  produce(id: ID!): Product
}

type User {
  email: String!
  name: String
  totalProductsCreated: Int
}

scalar _Any

union _Entity = Product | User

type _Service {
  \\"\\"\\"
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  \\"\\"\\"
  sdl: String
}

scalar link__Import"
`;
